zookeeper-server/src/main/java/org/apache/zookeeper/common/PathTrie.java

```diff
@@ -40,254 +46,275 @@
-         * get the list of children of this 
-         * trienode.
-         * @return the string list of its children
+         * Get the list of children of this trienode.
+         *
+         * @return A collection containing the node's children
          */
-        String[] getChildren() {
-           synchronized(children) {
-               return children.keySet().toArray(new String[0]);
-           }
+        Collection<String> getChildren() {
+            return children.keySet();
         }
-        
+
         /**
-         * get the string representation
-         * for this node
+         * Determine if this node is a leaf (has no children).
+         *
+         * @return true if this node is a lead node; otherwise false
          */
+        boolean isLeafNode() {
+            return children.isEmpty();
+        }
+
+        @Override
         public String toString() {
-            StringBuilder sb = new StringBuilder();
-            sb.append("Children of trienode: ");
-            synchronized(children) {
-                for (String str: children.keySet()) {
-                    sb.append(" " + str);
-                }
-            }
-            return sb.toString();
+          return "TrieNode [name=" + value + ", property=" + property
+              + ", children=" + children.keySet() + "]";
         }
     }
-    
+
     /**
-     * construct a new PathTrie with
-     * a root node of /
+     * Construct a new PathTrie with a root node.
      */
     public PathTrie() {
-        this.rootNode = new TrieNode(null);
+        this.rootNode = new TrieNode(null, "");
     }
-    
+
     /**
-     * add a path to the path trie 
-     * @param path
+     * Add a path to the path trie.
+     *
+     * @param path the path to add to the trie
      */
-    public void addPath(String path) {
-        if (path == null) {
-            return;
-        }
-        String[] pathComponents = path.split("/");
-        TrieNode parent = rootNode;
-        String part = null;
-        if (pathComponents.length <= 1) {
-            throw new IllegalArgumentException("Invalid path " + path);
+    public void addPath(final String path) {
+        Objects.requireNonNull(path, "Path cannot be null");
+
+        final String[] pathComponents = StringUtils.split(path, '/');
+        if (pathComponents.length == 0) {
+            throw new IllegalArgumentException("Invalid path: " + path);
         }
-        for (int i=1; i<pathComponents.length; i++) {
-            part = pathComponents[i];
-            if (parent.getChild(part) == null) {
-                parent.addChild(part, new TrieNode(parent));
+
+        writeLock.lock();
+        try {
+            TrieNode parent = rootNode;
+            for (final String part : pathComponents) {
+                TrieNode child = parent.getChild(part);
+                if (child == null) {
+                    child = new TrieNode(parent, part);
+                    parent.addChild(part, child);
```

Comment:
Let's take `addPath()` as an example.

In the previous version of `TrieNode` the class was already consistent, because methods were protected with synchronized and both getChild() and addChild() validated if the child was already added to `children`.
For instance, if a thread gets interrupted between getChild() and addChild() modifying the trie by adding the same child on another thread, the addChild() will be a no-op.

In your version `TrieNode` is no longer thread-safe, because you removed synchronized from all of its methods, but the entire tree is guarded by Read/Write lock in `PathTrie` class.

This seems to me heavier, because if multiple threads are manipulating different parts of the Trie, locking the entire object is not needed, because those threads are not conflicting. The previous version seems to me more accurate in certain workloads.